app/ota_client.py

```diff
@@ -314,6 +386,116 @@ def acquire_staging_storage(self):
             self._lock.release()
 
 
+class OtaStateSync:
+    """State machine that synchronzing ota_service and ota_client.
+
+    Typical usage:
+    a. wait on specific state
+        fsm: OtaStateSync
+        fsm.wait_on(fsm._S0, timeout=6)
+    b. expect <state>, and doing something to switch to next state
+        fsm: OtaStateSync
+        with fsm.proceed(fsm._P1, expect=fsm._START) as _next:
+            # do something here...
+            print(f"done! switch to next state {_next}")
+    """
+
+    ######## state machine definition ########
+    # states definition
+    _START, _S0, _S1, _S2, _END = (
+        "_START",  # start
+        "_S0",  # cache_scrub_finished
+        "_S1",  # pre_update_finished
+        "_S2",  # apply_update_finished
+        "_END",  # end
+    )
+    _STATE_LIST = ["_START", "_S0", "_S1", "_S2", "_END"]
+    # participators definition
+    _P1, _P2 = "ota_service", "ota_client"
+    # which participator can start the fsm
+    _STARTER = _P1
+
+    # input: (<expected_state>, <caller>)
+    # output: (<next_state>)
+    _STATE_SWITCH = {
```

Comment:
It seems difficult to read code because S0, S1, S2, P1, P2 variables have no meaning.
If P1 is like P_ota_service, I think it's OK.

Reply:
If developer needs details about the state machine, they can refer to the comments of `OtaStateSync` though. Anyway I will  rename P1, P2 to P_ota_service and P_ota_client accordingly.